File: sdks/node/src/brain.ts
import type { HttpClient } from "./client.js";
import type {
  BrainShareParams,
  BrainMemory,
  BrainSearchResult,
} from "./types.js";

const BRAIN_BASE_URL = "https://pi.ruv.io";

/** Interact with the Brain shared-knowledge system. */
export class BrainResource {
  constructor(private readonly client: HttpClient) {}

  /** Share a new memory / knowledge entry. */
  async share(params: BrainShareParams): Promise<BrainMemory> {
    return this.client.request<BrainMemory>("POST", "/brain/share", params);
  }

  /** Search the shared brain knowledge base. */
  async search(
    query: string,
    options?: { limit?: number; tags?: string[] },
  ): Promise<BrainSearchResult> {
    const params: Record<string, unknown> = { query };
    if (options?.limit) params.limit = options.limit;
    if (options?.tags) params.tags = options.tags;
    return this.client.request<BrainSearchResult>(
      "POST",
      "/brain/search",
      params,
    );
  }

  /** Vote on a brain memory entry (upvote / downvote). */
  async vote(
    memoryId: string,
    direction: "up" | "down",
  ): Promise<void> {
    await this.client.request<void>("POST", "/brain/vote", {
      memoryId,
      direction,
    });
  }
}

// Re-export for internal use — the BrainResource uses the main client but
// callers may need to configure the brain base URL separately in the future.
export { BRAIN_BASE_URL };

File: sdks/rust/src/brain.rs
use serde::Serialize;

use crate::client::Client;
use crate::error::Error;
use crate::types::BrainSearchResponse;

/// Operations on the brain / knowledge memory API.
pub struct BrainResource<'a> {
    pub(crate) client: &'a Client,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct ShareBody {
    content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    namespace: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tags: Option<Vec<String>>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct SearchBody {
    query: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    namespace: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    limit: Option<u32>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct VoteBody {
    memory_id: String,
    vote: i8,
}

/// Response from the share endpoint.
#[derive(Debug, Clone, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShareResponse {
    pub success: bool,
    #[serde(default)]
    pub id: Option<String>,
}

/// Response from the vote endpoint.
#[derive(Debug, Clone, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VoteResponse {
    pub success: bool,
}

impl<'a> BrainResource<'a> {
    /// Share a new memory / knowledge entry.
    pub async fn share(
        &self,
        content: &str,
        namespace: Option<&str>,
        tags: Option<Vec<String>>,
    ) -> Result<ShareResponse, Error> {
        let body = ShareBody {
            content: content.to_owned(),
            namespace: namespace.map(|s| s.to_owned()),
            tags,
        };
        self.client.post("/brainShare", &body).await
    }

    /// Semantic search across brain memories.
    pub async fn search(
        &self,
        query: &str,
        namespace: Option<&str>,
        limit: Option<u32>,
    ) -> Result<BrainSearchResponse, Error> {
        let body = SearchBody {
            query: query.to_owned(),
            namespace: namespace.map(|s| s.to_owned()),
            limit,
        };
        self.client.post("/brainSearch", &body).await
    }

    /// Vote on a memory entry (positive = 1, negative = -1).
    pub async fn vote(&self, memory_id: &str, vote: i8) -> Result<VoteResponse, Error> {
        let body = VoteBody {
            memory_id: memory_id.to_owned(),
            vote,
        };
        self.client.post("/brainVote", &body).await
    }
}
